metrics_metadata.go

```diff
@@ -0,0 +1,292 @@
+		message, _ := ioutil.ReadAll(resp.Body)
+		return nil, fmt.Errorf("Bad status %d: %s", resp.StatusCode, message)
+	}
+
+	finalDimension := &metrics_metadata.Dimension{}
+
+	err = json.NewDecoder(resp.Body).Decode(finalDimension)
+
+	return finalDimension, err
+}
+
+// UpdateDimension updates a dimension.
+func (c *Client) UpdateDimension(key string, value string, dim *metrics_metadata.Dimension) (*metrics_metadata.Dimension, error) {
+	payload, err := json.Marshal(dim)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.doRequest("PUT", DimensionAPIURL+"/"+key+"/"+value, nil, bytes.NewReader(payload))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		message, _ := ioutil.ReadAll(resp.Body)
+		return nil, fmt.Errorf("Bad status %d: %s", resp.StatusCode, message)
+	}
+
+	finalDimension := &metrics_metadata.Dimension{}
+
+	err = json.NewDecoder(resp.Body).Decode(finalDimension)
+
+	return finalDimension, err
+}
+
+// SearchDimension searches for dimensions, given a query string in `query`.
+func (c *Client) SearchDimension(query string, orderBy string, limit int, offset int) (*metrics_metadata.DimensionQueryResponseModel, error) {
+	params := url.Values{}
+	params.Add("query", query)
+	params.Add("orderBy", orderBy)
+	params.Add("limit", strconv.Itoa(limit))
+	params.Add("offset", strconv.Itoa(offset))
+
+	resp, err := c.doRequest("GET", DimensionAPIURL, params, nil)
+
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	finalDimensions := &metrics_metadata.DimensionQueryResponseModel{}
+
+	err = json.NewDecoder(resp.Body).Decode(finalDimensions)
+
+	return finalDimensions, err
+}
+
+// SearchMetric searches for metrics, given a query string in `query`.
+func (c *Client) SearchMetric(query string, orderBy string, limit int, offset int) (*metrics_metadata.RetrieveMetricMetadataResponseModel, error) {
+	params := url.Values{}
+	params.Add("query", query)
+	params.Add("orderBy", orderBy)
+	params.Add("limit", strconv.Itoa(limit))
+	params.Add("offset", strconv.Itoa(offset))
+
+	resp, err := c.doRequest("GET", MetricAPIURL, params, nil)
+
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	finalMetrics := &metrics_metadata.RetrieveMetricMetadataResponseModel{}
+
+	err = json.NewDecoder(resp.Body).Decode(finalMetrics)
+
+	return finalMetrics, err
+}
+
+func (c *Client) GetMetric(name string) (*metrics_metadata.Metric, error) {
+	resp, err := c.doRequest("GET", MetricAPIURL+"/"+name, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		message, _ := ioutil.ReadAll(resp.Body)
+		return nil, fmt.Errorf("Bad status %d: %s", resp.StatusCode, message)
+	}
+
+	finalMetric := &metrics_metadata.Metric{}
+
+	err = json.NewDecoder(resp.Body).Decode(finalMetric)
+
+	return finalMetric, err
+}
+
+func (c *Client) GetMetricTimeSeries(id string) (*metrics_metadata.MetricTimeSeries, error) {
```

Comment:
I need the ability to search MTSs by query (not id) for a Kubernetes HPA adapter I'm working on.  Would you be so kind as to throw this in this PR??  Otherwise I will add it.

Reply:
As part of my work, I'm probably going to also need a higher-level method that abstracts over the pagination that would probably be useful to put in this repo.  I'll write that though as part of the HPA adapter though and make a PR.

Reply:
Yes, I'll add that query bit. The higher level things sounds great!

Reply:
@keitwb will your HPA adapter be open source? Yelp would use it if you did that. cc @EmanekaT

Reply:
@solarkennedy @EmanekaT Yes, I went ahead and put a placeholder repo at https://github.com/signalfx/signalfx-k8s-metrics-adapter.  If you have any specific requirements, let me know, maybe by posting an issue on that repo.  Subscribe to releases if you want to know when the initial release is out.